Simple tree sitter based syntax highlighting for typescript

alexr00 opened the pull request:
Part of #210475

alexr00 reviewed (commented):
@hediet I think I got all the changes we discussed, with the exception of the change to `ITreeSitterTokenizationSupport` that we I will make in a subsequent PR (also as discussed). I made a new PR where my changes are squashed as there were some messy merge conflicts in the old one.

alexr00 commented on src/tsconfig.monaco.json:
```diff
@@ -8,7 +8,7 @@
 		"paths": {},
 		"module": "amd",
-		"moduleResolution": "classic",
+		"moduleResolution": "node",
```
@hediet is this ok to do?

hediet replied:
Honestly, I have no idea what this changes. Why did you have to change it?

alexr00 replied:
The "Monaco Editor checks" fail with this if I don't change it:

![image](https://github.com/user-attachments/assets/48add612-bbd1-48d4-9f85-385617037da8)

hediet commented on src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts:
```diff
@@ -54,10 +51,18 @@ export class TextModelTreeSitter extends Disposable {
 
 		const token = cancelOnDispose(this._languageSessionDisposables);
-		const language = await this._treeSitterParser.getLanguage(languageId);
+		let language = this._treeSitterLanguages.getLanguage(languageId);
+		if (!language) {
+			const languageAdded = Event.toPromise(Event.onceIf(this._treeSitterLanguages.onDidAddLanguage, (e) => e.id === languageId));
+			const canceled = new Promise<void>(resolve => token.onCancellationRequested(resolve));
+			const result = await Promise.race([languageAdded, canceled]);
+			if (result) {
+				language = result.language;
+			}
+		}
```
I think the pattern here would be `waitForDefinedState(getState: () => T | undefined, onDidChange: Event<unknown>, cancellationToken): Promise<T>`. The implementation would throw on cancellation and wait for `getState` to return a non-undefined value. It would check it initially and whenever the event fires.
`onceIf` can be dangerous, if the event never fires initially.

alexr00 replied:
I've changed to a `_getLanguage` method that throws on cancellation. When we're using the cancellation token to indicate that we should stop processing the language change, I do prefer the promise race though for the consistancy of only having one thing to check: has cancellation occurred.

hediet commented on src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts:
```diff
@@ -201,40 +209,56 @@ export class TreeSitterTree implements IDisposable {
 	}
 
-	public async getLanguage(languageId: string): Promise<Parser.Language | undefined> {
+	public getLanguage(languageId: string): Parser.Language | undefined {
```
I think it is surprising that `getLanguage` causes a side effect (triggering to load the language), but still returns `undefined`.
Can't you return `Promise<Parser.Language>` here? That would also simplify the call-site.

alexr00 replied:
I need it to be sync. I'll rename it.

hediet commented on src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts:
```diff
@@ -264,9 +288,11 @@ export class TreeSitterTextModelService extends Disposable implements ITreeSitte
 	private _init!: Promise<boolean>;
 	private _textModelTreeSitters: DisposableMap<ITextModel, TextModelTreeSitter> = this._register(new DisposableMap());
-	private _registeredLanguages: DisposableMap<string, IDisposable> = this._register(new DisposableMap());
+	private readonly _registeredLanguages: Map<string, string> = new Map();
```
(feel free to ignore)

```suggestion
	private readonly _registeredLanguages = new Map<string, string>();
```

hediet commented on src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts:
```diff
@@ -54,11 +51,14 @@ export class TextModelTreeSitter extends Disposable {
+		try {
+			language = await this._getLanguage(languageId, token);
+		} catch (e) {
+			// cancellation has been requested.
```
```suggestion
			if (isCancellationError(e)) { return; }
			throw e;
```

alexr00 commented on src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts:
```diff
@@ -201,40 +222,56 @@ export class TreeSitterTree implements IDisposable {
 
 export class TreeSitterLanguages extends Disposable {
 	private _languages: Map<string, Parser.Language> = new Map();
+	protected /*exposed for tests*/ readonly _onDidAddLanguage: Emitter<{ id: string; language: Parser.Language }> = this._register(new Emitter());
```
Thanks for the reminder!

alexr00 commented on src/vs/workbench/services/treeSitter/browser/treeSitterTokenizationFeature.ts:
```diff
@@ -0,0 +1,241 @@
+		const lineLength = textModel.getLineMaxColumn(lineNumber);
+		const captures = this._captureAtRange(lineNumber, new ColumnRange(1, lineLength), textModel);
+
+		if (captures.length === 0 && lineLength > 0) {
```
It is supposed to be for empty lines to indicate that we have not tokens.

hediet replied:
Then I'd just check `if (lineLength === 0) { return undefined; }`. (make sure that lineLength has the correct value)

alexr00 replied:
Ah, sorry, it's for if we have a non-empty line, but there are no captures. I think before it returned an empty array, but you preferred undefined.

alexr00 replied:
Removed the check.

alexr00 commented on src/vs/workbench/contrib/codeEditor/browser/inspectEditorTokens/inspectEditorTokens.ts:
```diff
@@ -603,6 +648,28 @@ class InspectEditorTokensWidget extends Disposable implements IContentWidget {
 		return null;
 	}
 
+	private _walkTreeforPosition(cursor: Parser.TreeCursor, pos: Position): Parser.SyntaxNode | null {
+		const offset = this._model.getOffsetAt(pos);
+		cursor.gotoFirstChild();
+		let goChild: boolean = false;
+		let lastGoodNode: Parser.SyntaxNode | null = null;
+		do {
+			if (cursor.currentNode.startIndex <= offset && offset < cursor.currentNode.endIndex) {
+				goChild = true;
+				lastGoodNode = cursor.currentNode;
+			} else {
+				goChild = false;
+			}
+		} while (goChild ? cursor.gotoFirstChild() : cursor.gotoNextSibling());
+		return lastGoodNode;
+	}
+
+	private _getTreeSitterTokenAtPosition(tree: Parser.Tree, pos: Position): Parser.SyntaxNode | null {
+		const cursor = tree.walk();
+
+		return this._walkTreeforPosition(cursor, pos);
+	}
```
Will chang later.

alexr00 commented on src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts:
```diff
@@ -201,40 +222,56 @@ export class TreeSitterTree implements IDisposable {
+		}
+	}
+
+	private async _addLanguage(languageId: string): Promise<void> {
 		let language = this._languages.get(languageId);
 		if (!language) {
 			language = await this._fetchLanguage(languageId);
```
Added the Async Cache.

alexr00 commented on src/vs/workbench/services/treeSitter/browser/treeSitterTokenizationFeature.ts:
```diff
@@ -0,0 +1,241 @@
+	 * @returns
+	 */
+	public tokenizeEncoded(lineNumber: number, textModel: ITextModel): Uint32Array | undefined {
+		const lineLength = textModel.getLineMaxColumn(lineNumber);
```
I think I do want the column, not the length.

alexr00 commented on src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts:
```diff
@@ -67,17 +67,35 @@ export class TextModelTreeSitter extends Disposable {
+		if (language) {
+			return Promise.resolve(language);
+		}
+		return new Promise((resolve, reject) => {
+			this._languageSessionDisposables.add(this._treeSitterLanguages.onDidAddLanguage(e => {
+				if (e.id === languageId) {
+					resolve(e.language);
+				}
+			}));
+			const disposables: IDisposable[] = [];
+			this._languageSessionDisposables.add({ dispose: () => dispose(disposables) });
+			token.onCancellationRequested(reject, undefined, disposables);
+		});
```
I didn't see anything like `waitForEvent`. I've modified it slightly.

alexr00 commented on src/vs/editor/common/services/treeSitterParserService.ts:
```diff
@@ -3,14 +3,23 @@
- * Currently this service just logs telemetry about how long it takes to parse files.
- * Actual API will come later as we add features like syntax highlighting.
- */
 export interface ITreeSitterParserService {
 	readonly _serviceBrand: undefined;
+	onDidAddLanguage: Event<{ id: string; language: Parser.Language }>;
+	getOrInitLanguage(languageId: string): Parser.Language | undefined;
+	getParseResult(textModel: ITextModel): ITreeSitterParseResult | undefined;
+}
```
Leaving as is for now.

alexr00 commented on src/vs/workbench/services/treeSitter/browser/treeSitterTokenizationFeature.ts:
```diff
@@ -0,0 +1,241 @@
+	 * Gets the tokens for a given line.
+	 * Each token takes 2 elements in the array. The first element is the offset of the end of the token *in the line, not in the document*, and the second element is the metadata.
+	 *
+	 * @param lineNumber
```
test

alexr00 commented:
@hediet thanks for the review! I'll continue addressing your comments in the next PR, but this one was quite big and I don't want to merge it any later in the iteration.
